### PR TITLE
Submission: OrthoInit + Int6 MLP3x + SmearGate + BigramHash (val_bpb: 1.1524)

### PR DESCRIPTION
**Pull request by jfprincz**
## Submission: OrthoInit + Int6 MLP3x + SmearGate + BigramHash (val_bpb: 1.1524)

**val_bpb: 1.1524** (sliding window, stride=256) | **15.4 MB** | 8xH100 SXM, 600s

### Progress from our prior submission

| | PR #70 (v1) | This (v2) | Delta |
|---|---|---|---|
| **val_bpb (sliding)** | 1.1659 | **1.1524** | **-0.0135** |
| Artifact | 14.9 MB | 15.4 MB | +0.5 MB |
| Steps (600s) | 12,485 | 8,390 | -4,095 |
| Step time (8xH100) | 48ms | 68ms | +20ms |
| Train seq_len | 1024 | 2048 | 2x |
| Model params | 21.8M | 22.4M | +0.6M |

Fewer steps at longer context and richer initialization more than compensate for the slower per-step speed. The 0.0135 BPB improvement comes from stacking eight techniques that each contribute independently.

### Techniques

| # | Technique | Description |
|---|-----------|-------------|
| 1 | **Orthogonal + muP init** | `orthogonal_` on all large matrices, output projections scaled by 1/√(2·layers) — faster early convergence |
| 2 | **3x MLP** (hidden=1536) | +2.3M params, ~0.02 BPB gain — budget from int6 savings |
| 3 | **Int6 mixed quant + zstd-22** | Per-row int6 on MLP+attention, int8 on embeddings+bigram, fp32 on controls |
| 4 | **SmearGate** | Learned sigmoid gate blending each token with previous token's embedding (~512 params) |
| 5 | **Bigram Hash Embedding** | 4096-bucket hash table (dim=128→512) injecting token-pair features |
| 6 | **Tuned Muon optimizer** | LR=0.02, momentum=0.99 w/ warmup, warmdown=3000, grad clip=0.3 |
| 7 | **Seq 2048 + sliding window** | Train and eval at 2048 tokens, NTK-aware RoPE, stride=256 scoring |
| 8 | **FlashAttention 3** | Direct `flash_attn_func` calls — ~5% step time reduction vs SDPA backend |

### Results

| Metric | Value |
|--------|-------|
| Pre-quant val_bpb | 1.1659 |
| Int6 roundtrip val_bpb | 1.1744 |
| **Int6 sliding val_bpb (s256)** | **1.1524** |
| Steps completed (600s cap) | 8,390 |
| Step time | 68ms |
| Model params | 22,368,841 |
| Artifact size | 15,401,594 bytes |

### Reproducibility (3 seeds)

| Seed | Steps | Sliding s256 | Sliding s64 | Artifact |
|------|-------|-------------|-------------|----------|
| 1337 | 7,569 | 1.1544 | — | 15.40 MB |
| **42** | **8,390** | **1.1524** | **1.1525** | **15.40 MB** |
| 2025 | 8,712 | 1.1546 | 1.1546 | 15.47 MB |

Mean: **1.1538** | Variance: 0.0022 | Submitted: seed 42

### Run command

```bash
ITERATIONS=9000 MAX_WALLCLOCK_SECONDS=600 \
MATRIX_LR=0.02 SCALAR_LR=0.02 TIED_EMBED_LR=0.03 \
MUON_MOMENTUM=0.99 MUON_MOMENTUM_WARMUP_START=0.92 \
MUON_MOMENTUM_WARMUP_STEPS=1500 WARMDOWN_ITERS=3000 \
torchrun --standalone --nproc_per_node=8 train_gpt.py
```

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Submission: OrthoInit + Int6 MLP3x + SmearGate + BigramHash (val_bpb: 1.1524)

**BPB:** 1.1524 | **Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

**What I found in the code** (head SHA `37e6f3983456`, file `records/track_10min_16mb/2026-03-20_OrthoInit_Muon_Int6_MLP3x_SmearBigram_1.1524/train_gpt.py`):

Static code review found no TTT adaptation function, no SLOT optimization loop, no n-gram-cache class, and no pre-quant val-token fine-tune. The eval path uses the standard sliding-window stride-64 pattern. The submission is a pure-neural architecture iteration on the standard SP1024/SP4096/SP8192 baseline.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 0.03s, dim=512, layers=9, vocab=1024, code=64123 B, SMOKE_TEST_PASS

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the classification pass — this looks like a clean pure-neural iteration on the standard baseline.

*Auto-classification caveat:* this review was drafted by the AST-based classifier. If there's a non-standard eval mechanism (logit postprocessing, hedge mixing, etc.) that I missed because it's factored into a helper file or a non-standard function name, please flag it and I'll re-run the audit manually.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 0.03s, dim=512, layers=9, vocab=1024, code=64123 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*
